### DISCUSSION TITLE
Does sftpgo OIDC integartion automatically create Users in sftpgo?

### DISCUSSION THREAD

**Discussion by japatel** [drakkan/sftpgo#1685] [Q&A]
Does sftpgo OIDC integartion automatically create Users in sftpgo? 

Also, if the OIDC has groups which can be sent as claims. Can OIDC enabled sftpgo map the groups to this newly created account?

While connecting to sftpgo using standard Winscp, or sftp client, can sftpgo be configured to use OIDC for these connetions?

**Comment by fostermi** [comment]
The Event Manager can be used for this. Create an Action of type "Identity Provider Account Check" and use a template something like this: 
```
{
  "status": 1,
  "username": "{{Name}}",
  "homedir": "",
  "permissions": {
    "/": [
      "*"
    ]
  },
  "groups": [
      {
         "name": "group1",
          "type": 1
      }
   ]
}
```
Then create and Event Action Rule with Trigger "Identity Provider Logins" and Type "User Login" and use the action created above. 

I have not figured out a way to get group mappings from an OIDC "groups" claim though, which is my next stumbling block. I hope someone has figured that out.

**Reply by keriati** [reply]
Thanks, I got this working.
Is there somewhere a reference about this json structure in any documentation? I just failed to find anything so far.

**Comment by fostermi** [comment]
So I've figured out how to dynamically add users to groups from an OIDC custom claim of "groups" using a modified script as a "pre-login hook" as described in https://github.com/drakkan/sftpgo/discussions/1072.

The gist of it is to add a key of `"groups": $groups` in the `SFTPGO_USER_TEMPLATE` in the script in that discussion and then add the following lines to extract the list of groups from the json user representation passed to the script:

```
# Create a list of groups, each with new attribute "type". These will be secondary groups (2).
CURR_USER_GROUPS=$(echo "${SFTPGO_LOGIND_USER}" | jq -r '[.oidc_custom_fields.groups[] | select(.|test("^sftpgo-")) | { "type": 2, "name": . } ] ')

# Create the list of SFTGO group by appending the primary 'sftpgo-users' group (type 1)
SFTPGO_USER_GROUPS=$(echo "${CURR_USER_GROUPS}" | jq  '. + [{ "type": 1, "name": "sftpgo-users" }]')
```

Finally, adjust the `USER_TO_CREATE` variable assignment to replace the placeholder `groups` in the user template with the extracted list of user groups.
```
--argjson groups "${SFTPGO_USER_GROUPS}" \
```

In my example, I automatically add every user created to the group `sftpgo-users` and make that their primary group.  All other groups are prefixed by `sftpgo-` and I assign those groups as secondary groups. These groups need to exist in `sftpgo` beforehand.

There might be a better way to use `jq` to do this as a one-liner, but I couldn't figure out how to do that so its two lines to create the correct groups list.

Thanks to the original discussion author "shaltz" and drakkan for doing the majority of the work here. Hope this helps someone else.

**Reply by tecosaur** [reply]
I think this will likely help me, thanks for sharing!

**Comment by UtkarshVerma** [comment]
I ended up using a pre-login-hook instead since it does not rely on mouse-fu. Leaving it here for others:

```sh
#!/bin/sh
# Create a user on OIDC login if it doesn't exist.
#
# https://docs.sftpgo.com/2.6/dynamic-user-mod/

# Extract fields from user JSON.
# USER_ID=$(echo "$SFTPGO_LOGIND_USER" | jq -r '.id')
# USERNAME=$(echo "$SFTPGO_LOGIND_USER" | jq -r '.username')

USER_ID=$(echo "$SFTPGO_LOGIND_USER" | sed -n 's/.*"id":[[:space:]]*\([0-9]*\).*/\1/p')
USERNAME=$(echo "$SFTPGO_LOGIND_USER" | sed -n 's/.*"username":[[:space:]]*"\([^"]*\)".*/\1/p')

# Only proceed if the user doesn't exist.
[ "$USER_ID" != 0 ] && exit

# Only proceed if login method is OIDC.
[ "$SFTPGO_LOGIND_METHOD" != "IDP" ] && exit
[ "$SFTPGO_LOGIND_PROTOCOL" != "OIDC" ] && exit

# Write the user to stdout to create the user.
# https://github.com/drakkan/sftpgo/blob/2.6.x/openapi/openapi.yaml#L5847
cat <<EOF
{
    "status": 1,
    "username": "$USERNAME",
    "has_password": false,
    "permissions": {
        "/": ["*"]
    }
}
EOF
```

Make sure to point the `SFTPGO_DATA_PROVIDER__PRE_LOGIN_HOOK` env var to your hook and that's all. This won't work for distroless variants of the Docker container by the way since you'll need the shell for running the hook.
